mater/lib/src/file_reader.rs

```diff
@@ -0,0 +1,246 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    io::Cursor,
+    ops::Deref,
+    path::Path,
+};
+
+use async_stream::try_stream;
+use futures::{Stream, TryStreamExt};
+use ipld_core::{cid::Cid, codec::Codec};
+use ipld_dagpb::{DagPbCodec, PbNode};
+use tokio::{
+    fs::File,
+    io::{AsyncRead, AsyncSeek, AsyncSeekExt, AsyncWriteExt},
+};
+
+use crate::{multicodec, v1::BlockMetadata, v2, Error};
+
+/// CAR file loader.
+pub struct FileLoader<R>
+where
+    R: AsyncRead + AsyncSeek + Unpin,
+{
+    reader: v2::Reader<R>,
+    index: HashMap<Cid, PartialNode>,
+}
+
+impl FileLoader<File> {
+    /// Creates a [`FileLoader`] from the given file path.
+    pub async fn from_path<P>(path: P) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = File::open(path).await?;
+        let mut loader = Self {
+            reader: v2::Reader::new(file),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
+        Ok(loader)
+    }
+}
+
+impl FileLoader<Cursor<Vec<u8>>> {
+    /// Creates a [`FileLoader`] from a vector of bytes.
+    pub async fn from_vec(vec: Vec<u8>) -> Result<Self, Error> {
+        let mut loader = Self {
+            reader: v2::Reader::new(Cursor::new(vec)),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
+        Ok(loader)
+    }
+}
+
+impl<R> FileLoader<R>
+where
+    R: AsyncRead + AsyncSeek + Unpin,
+{
+    /// Returns the root of the CAR file.
+    ///
+    /// If the number of roots is not 1, returns [`Error::WrongNumberOfRoots`].
+    pub async fn root(&mut self) -> Result<Cid, Error> {
+        self.reader.get_inner_mut().rewind().await?;
+        self.reader.read_pragma().await?;
+        self.reader.read_header().await?;
+        // We assume there's a single root
+        self.reader
+            .read_v1_header()
+            .await?
+            .roots
+            .first()
+            .copied()
+            .ok_or(Error::WrongNumberOfRoots)
+    }
+
+    /// Indexes the file.
+    ///
+    /// *Always* seeks to the start of the file before proceeding with indexing.
+    /// This function performs indexing *naively*, it will sequentially read all block "headers"
+    /// (Cid, data start offset and data length) without loading the blocks into memory.
+    async fn index(&mut self) -> Result<(), Error> {
```

Comment:
nit: judging by private field `index` and this function name, I'd say that this function returns private field `index`.
```suggestion
    async fn create_index(&mut self) -> Result<(), Error> {
```